apps/extension/src/core/domains/chains/handler.ts

```diff
@@ -0,0 +1,230 @@
+import { verifierCertificateMnemonicStore } from "@core/domains/accounts/store.verifierCertificateMnemonic"
+import { talismanAnalytics } from "@core/libs/Analytics"
+import { ExtensionHandler } from "@core/libs/Handler"
+import { generateQrAddNetworkSpecs, generateQrUpdateNetworkMetadata } from "@core/libs/QrGenerator"
+import { chaindataProvider } from "@core/rpcs/chaindata"
+import { WsProvider } from "@polkadot/api"
+import { assert, u8aToHex } from "@polkadot/util"
+import { CustomSubNativeToken, subNativeTokenId } from "@talismn/balances-substrate-native"
+import {
+  BalanceMetadata,
+  CustomChain,
+  githubUnknownTokenLogoUrl,
+} from "@talismn/chaindata-provider"
+import {
+  $metadataV14,
+  PalletMV14,
+  StorageEntryMV14,
+  filterMetadataPalletsAndItems,
+} from "@talismn/scale"
+import * as $ from "@talismn/subshape-fork"
+import { sleep } from "@talismn/util"
+import { MessageHandler, MessageTypes, RequestType, RequestTypes, ResponseType } from "core/types"
+
+export class ChainsHandler extends ExtensionHandler {
+  private chainUpsert: MessageHandler<"pri(chains.upsert)"> = async (chain) => {
+    let customBalanceMetadata: BalanceMetadata[] | undefined = undefined
+    if (chain.id === `custom-${chain.genesisHash}`) {
+      // When saving custom chains, download the chain metadata and build some SCALE types so we can fetch balances.
+      //
+      // TODO: Replace this with a scheduled background task which maintains (keeps up to date) the required SCALE types for *all* chains, not just custom ones.
+
+      const rpcUrl = chain.rpcs?.find((rpc) => rpc.url && /^wss?:\/\/.+$/.test(rpc.url))
+      if (!rpcUrl) throw new Error("No valid RPC found")
+
+      const ws = new WsProvider(rpcUrl.url, 0)
+      const [
+        // TODO: Store minMetadata with genesisHash|specName|specVersion index for scheduled background task
+        // genesisHash,
+        // { specName, specVersion },
+        metadataRpc,
+      ] = await (async () => {
+        try {
+          await ws.connect()
+
+          const isReadyTimeout = sleep(10_000).then(() => {
+            throw new Error("timeout")
+          })
+          await Promise.race([ws.isReady, isReadyTimeout])
+
+          return await Promise.all([
+            // TODO: Store minMetadata with genesisHash|specName|specVersion index for scheduled background task
+            // ws.send<string>("chain_getBlockHash", [0]),
+            // ws.send<{ specName: string; specVersion: number }>("state_getRuntimeVersion", []),
+            ws.send<string>("state_getMetadata", []),
```

Comment:
I think metadata (or minMetadata here) should always be saved along with specName/specVersion/genesisHash so it can be updated if any of these change.
This feature is for devs so they may run different chains on same port from one day to another, need to be able to reset the metadata if anything is suspicious.

Reply:
I agree in principle, but instead of spending more time on this as part of this PR I would prefer to do it as part of https://github.com/TalismanSociety/talisman/issues/1076.

I'm already working on that issue, and the logic in that issue will work for *all* chains, not just custom ones.

So basically the metadata handling in this PR is only going to be used for one release of the wallet.  
Then in the next release it'll be deleted.

Reply:
gotcha, perfect